File: Top200/offer36-TreeToDoublyList.java
package Top200;

class TreeToDoublyList {
    class Node {
        public int val;
        public Node left;
        public Node right;

        public Node() {}

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val,Node _left,Node _right) {
            val = _val;
            left = _left;
            right = _right;
        }
    }

    Node pre = null;
    Node head = null;
    
    public Node treeToDoublyList(Node root) {
        if (root == null) {
            return null;
        }
        
        dfs(root);

        pre.right = this.head;
        this.head.left = pre;
        return head;
    }
    
    public void dfs(Node root) {
        if (root == null) {
            return;
        }

        dfs(root.left);

        if (this.pre == null) {
            this.head = root;
        } else {
            this.pre.right = root;
        }
        
        root.left = this.pre; 
        this.pre = root;

        dfs(root.right);
    }
}

File: Top200/offer36-TreeToDoublyList.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var (
	prev *TreeNode = nil
	head *TreeNode = nil
)

// 20210826: LeetCode 上没有 Go 语言这个选项
// 所以以下代码暂时未验证
func treeToDoublyList(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	head.Left = prev
	prev.Right = head

	return head
}

func toDoublyList(root *TreeNode) {
	if root == nil {
		return
	}

	toDoublyList(root.Left)

	if prev == nil {
		head = root
	} else {
		prev.Right = root
	}

	root.Left = prev
	prev = root

	toDoublyList(root.Right)
}
